File: src/3494.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def minTime(self, skill: List[int], mana: List[int]) -> int:
        retVal = 0

        skillSize = len(skill)
        manaSize = len(mana)

        times = [0] * skillSize
        for j in range(manaSize):
            currentTime = 0
            for i in range(skillSize):
                currentTime = max(currentTime, times[i]) + skill[i] * mana[j]
            times[skillSize - 1] = currentTime
            for i in range(skillSize - 2, -1, -1):
                times[i] = times[i + 1] - skill[i + 1] * mana[j]
        retVal = times[skillSize - 1]

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for skill, mana in [[[1, 5, 2, 4], [5, 1, 4, 2]], [[1, 1, 1], [1, 1, 1]], [[1, 2, 3, 4], [1, 2]]]:
            # /* Example
            #  *  Input: skill = [1,5,2,4], mana = [5,1,4,2]
            #  *  Output: 110
            #  *
            #  *  Input: skill = [1,1,1], mana = [1,1,1]
            #  *  Output: 5
            #  *
            #  *  Input: skill = [1,2,3,4], mana = [1,2]
            #  *  Output: 21
            #  */
            logging.info("Input: skill = %s, mana = %s", skill, mana)

            retVal = pSolution.minTime(skill, mana)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)

File: src/3494.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

long long minTime(int* skill, int skillSize, int* mana, int manaSize) {
    long long retVal = 0;

    long long times[skillSize];
    memset(times, 0, sizeof(times));
    long long currentTime;
    for (int j = 0; j < manaSize; j++) {
        currentTime = 0;
        for (int i = 0; i < skillSize; i++) {
            currentTime = ((currentTime > times[i]) ? (currentTime) : (times[i])) + (long long)(skill[i] * mana[j]);
        }
        times[skillSize - 1] = currentTime;
        for (int i = skillSize - 2; i >= 0; i--) {
            times[i] = times[i + 1] - (long long)skill[i + 1] * mana[j];
        }
    }
    retVal = times[skillSize - 1];

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (5000)
    struct testCaseType {
        int skill[MAX_SIZE];
        int skillSize;
        int mana[MAX_SIZE];
        int manaSize;
    } testCase[] = {{{1, 5, 2, 4}, 4, {5, 1, 4, 2}, 4}, {{1, 1, 1}, 3, {1, 1, 1}, 3}, {{1, 2, 3, 4}, 4, {1, 2}, 2}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: skill = [1,5,2,4], mana = [5,1,4,2]
     *  Output: 110
     *
     *  Input: skill = [1,1,1], mana = [1,1,1]
     *  Output: 5
     *
     *  Input: skill = [1,2,3,4], mana = [1,2]
     *  Output: 21
     */

    long long answer;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: skill = [");
        for (j = 0; j < testCase[i].skillSize; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", testCase[i].skill[j]);
        }
        printf("], mana = [");
        for (j = 0; j < testCase[i].manaSize; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", testCase[i].mana[j]);
        }
        printf("]\n");

        answer = minTime(testCase[i].skill, testCase[i].skillSize, testCase[i].mana, testCase[i].manaSize);
        printf("Output: %lld\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}
